src/ucp/core/ucp_worker.c

```diff
@@ -2445,8 +2448,18 @@ static void ucp_worker_discard_uct_ep_cleanup(ucp_worker_h worker)
 
     kh_foreach(&worker->discard_uct_ep_hash, uct_ep, req, {
         ucs_assert(uct_ep == req->send.discard_uct_ep.uct_ep);
-        uct_ep_pending_purge(uct_ep, ucp_worker_discard_uct_ep_purge, NULL);
+        purged_req = NULL;
+        uct_ep_pending_purge(uct_ep, ucp_worker_discard_uct_ep_purge,
+                             &purged_req);
         uct_ep_destroy(uct_ep);
+        if (purged_req != NULL) {
+            /* Since uct_ep_destroy() may complete some in-flight operations,
+             * UCP endpoint must stay valid, so complete discard operation and
+             * remove reference counter of UCP endpoint after destroying of UCT
+             * endpoint */
+            ucs_assert(req == purged_req);
+            ucp_worker_discard_uct_ep_complete(req);
+        }
 
         /* Either flush is in-progress (will be completed in UCT EP destroy)
```

Comment:
`/* We must do this operation as last step, because uct_ep_destroy could move a discard operation to the progress queue */`

Reply:
done